### PR TITLE
Reregister TcpStream on partial reads/write on Windows

### LINKED ISSUE #1078: Investigate TcpStream smoke test failures
Pr #1070 is going to add (or added, depending on when you read this) two smoke tests for  `TcpStream` `tcp_stream_ipv4` and `tcp_stream_ipv6`, which only fails on Windows. Figure out why.

/cc @PerfectLaugh

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
To Emulate edge-triggers.

Closes #1078.

### CODE REVIEW

File: src/sys/windows/mod.rs [hunk 1]
```diff
@@ -34,6 +34,19 @@ macro_rules! try_io {
         }
         result
     }};
+    ($self: ident.$method: ident ( $($arg: expr),* $(,)* ), $reregister_check: expr)  => {{
+        let result = (&$self.inner).$method($($arg),*);
+        let should_register = match &result {
```
Comment (dtacalau) [inline src/sys/windows/mod.rs]: I'd rename this to should_reregister.

Comment (dtacalau) [reply src/sys/windows/mod.rs]: Good job, nice to see this working.

Comment (Thomasdezeeuw) [reply src/sys/windows/mod.rs]: Done.

**Comment by dtacalau** [comment]
Just a notice: with this change test peek from tcp.rs blocks
```
tests/tcp.rs
test peek ... test peek has been running for over 60 seconds
```

Haven't looked into it yet.

**Comment by Thomasdezeeuw** [comment]
Don't merge this yet, I'm enabling the peek test again see if it works this time.

**Comment by Thomasdezeeuw** [comment]
Ready to be merged now. The peek test passed so it seemed like some fluke, or it was fixed in the mean time.

**Comment by dtacalau** [comment]
@Thomasdezeeuw what's holding you back from merging this?

**Pull request merged**